app/services/sync_ndbn_members.rb

```diff
@@ -6,26 +6,31 @@ def self.upload(member_file)
       return [e.message]
     end
 
-    member_entries.flat_map do |member_id, account_name|
-      ndbn_member = NDBNMember.find_or_initialize_by(ndbn_member_id: member_id.to_i)
+    member_entries.flat_map do |row|
+      member_id, account_name = row[:ndbn_member_number], row[:member_name]
+      ndbn_member = NDBNMember.find_or_initialize_by(ndbn_member_id: member_id)
       ndbn_member.account_name = account_name
       ndbn_member.save
 
+      begin
+        Integer(member_id)
+      rescue ArgumentError
+        ndbn_member.errors.add(:ndbn_member_id, "id must be an integer")
+      end
+
       ndbn_member.errors.full_messages.flat_map do |msg|
-        "Issue with #{member_id}: #{account_name} -> #{msg}"
+        "Issue with '#{member_id},#{account_name}'-> #{msg}"
       end
     end
   end
 
   def self.parse_csv(member_file)
     raise ParseError, "CSV upload is required." if member_file.nil?
 
-    raw = CSV.parse(member_file)
-
-    raw.select do |member_id, member_name|
-      member_id.match(/^\d+$/) && member_name.present?
-    end
-  rescue CSV::MalformedCSVError
+    data = member_file.readlines
+    data = data.drop(1) if /Update/.match?(data.first)
+    CSV.parse(data.join, headers: true, header_converters: :symbol)
+  rescue ArgumentError
```

Comment:
Not sure I understand what's going on here?

Reply:
basically, the CSV starts with `Updated: 23/132/123,` I need to remove that line to have the CSV parsing work. The second line is headers. 

The way I see it I have two options:
- strip the first line, parse the second as headers 
  - from what I can tell, I can't tell CSV to skip the first row. If I want to use the headers I need to change the CSV that is passed to the `CSV.parse` to not have that first row. So I open the file, drop the first row if its bad then pipe to CSV.
- strip both lines, use the data as an array of arrays

the second option is probably more clear so how about?
```rb
  def self.parse_csv(member_file)
    raise ParseError, "CSV upload is required." if member_file.nil?

    data = CSV.parse(member_file)
    # Remove the first row because csv starts with Updated: 1/10/2024,
    data = data.drop(1) if data.first&.first&.start_with?("Updated")
    # Remove the headers, we are not using them.
    data = data.drop(1) if data.first&.first&.match?(/NDBN/i)
  rescue ArgumentError
    raise ParseError, "The CSV File provided was invalid."
  end

```

Reply:
Hmm... not the biggest fan of this. Since it's only super users doing this, I'd rather force them to remove the first row manually before uploading. Then we can parse using CSV's `headers` parameter and everything "just works" a lot more flexibly.

We can add instructions to the page indicating that they need to ensure what the first and subsequent rows look like (by showing a sample).

@cielf thoughts?

Reply:
Well, I know I'm often somewhat distracted  -- what would the consequences be in this case if someone neglected to remove the line.

Reply:
The headers would get messed up => every item gets parsed wrong => you get a 500 cookie overflow error because the session cookie runs out of space to show all the errors. 

The last part can be fixed by limiting the amount of errors shown but largely the result will be a bunch of errors shown like:

![image](https://github.com/rubyforgood/human-essentials/assets/14540596/9a73edcc-df9c-4225-b1fc-ca6b0dcc651a)

Reply:
Could do something like 
![image](https://github.com/rubyforgood/human-essentials/assets/14540596/6ecf1710-f8b7-4f87-8d1e-aed7f42d161d)

Reply:
Can we not validate that the first line of the CSV looks the way we expect? Using something like this [https://dpericich.medium.com/how-to-use-ruby-to-get-csv-headers-3f3bffd3e965](https://dpericich.medium.com/how-to-use-ruby-to-get-csv-headers-3f3bffd3e965) to get the headers and immediately raise an error if they don't look the way we expect.

Reply:
![image](https://github.com/rubyforgood/human-essentials/assets/14540596/f2e4c39a-f831-4ba2-9bf9-268b059bf311)

Man, this is why I love code review. We can certainly do that. I feel very stupid :upside_down_face:. Definitely tunneled a bit on my solution. Appreciate the more clear headed take.

Reply:
*nods* I think that will do for now -- in the longer term I'd still rather foist it off on NDBN, of course.